src/atomic.rs

```diff
@@ -1,160 +1,380 @@
-        Ok(cbor::deserialize_params(ser)?)
+    pub fn load(cid: &Cid, bs: &impl Blockstore) -> anyhow::Result<Option<AtomicExecRegistry>> {
+        bs.get_cbor(cid)
     }
-}
 
-/// State of an actor including a lock to support atomic executions.
-#[derive(Serialize, Deserialize_tuple)]
-pub struct LockableState<T>
-where
-    T: Serialize + DeserializeOwned + MergeableState<T>,
-{
-    lock: bool,
-    state: T,
+    pub fn init_atomic_exec<'a, S>(
+        &mut self,
+        bs: &impl Blockstore,
+        state: impl IntoIterator<Item = &'a mut S>,
+        input: AtomicInput,
+        lock: bool,
+    ) -> anyhow::Result<AtomicInputID>
+    where
+        S: LockableState + 'a,
+    {
+        // Optionally lock the state and compute its CIDs
+        let unlocked_state_cids = state.into_iter().try_fold(Vec::new(), |mut v, s| {
+            if lock {
+                s.lock()?;
+            } else {
+                v.push(s.cid());
+            }
+            anyhow::Ok(v)
+        })?;
+
+        // Generate and register a new input ID
+        let input_id = self.new_input_id(&unlocked_state_cids, &input);
+        self.input_ids.modify(bs, |m| {
+            let k = BytesKey::from(input_id.bytes());
+            let v = m.set(
+                k,
+                AtomicInputEntry {
+                    unlocked_state_cids,
+                    input,
+                },
+            )?;
+            assert!(v.is_some(), "input ID collision");
+            Ok(())
+        })?;
+
+        Ok(input_id)
+    }
+
+    pub fn cancel_atomic_exec<'a, S>(
+        &mut self,
+        bs: &impl Blockstore,
+        input_id: AtomicInputID,
+        input_fn: impl FnOnce(AtomicInput) -> Box<dyn Iterator<Item = &'a mut S>>,
+    ) -> anyhow::Result<()>
+    where
+        S: LockableState + 'a,
+    {
+        // Consume own input ID and retrieve the associated data
+        let AtomicInputEntry { input, .. } = self.input_ids.modify(bs, |m| {
+            let k = BytesKey::from(input_id.bytes());
+            let (_, v) = m
+                .delete(&k)?
+                .ok_or_else(|| anyhow::anyhow!("unexpected input ID"))?;
+            Ok(v)
+        })?;
+
+        // Get the state and ensure it's unlocked
+        let state_iter = input_fn(input);
+        state_iter.for_each(|s| {
+            if s.is_locked() {
+                s.unlock().unwrap();
+            }
+        });
+
+        Ok(())
+    }
+
+    pub fn prepare_atomic_exec<'a, S, I>(
+        &mut self,
+        bs: &impl Blockstore,
+        own_input_id: AtomicInputID,
+        input_ids: &HashMap<IPCAddress, AtomicInputID>,
+        input_fn: impl FnOnce(AtomicInput) -> (I, Box<dyn Iterator<Item = &'a mut S>>),
+        output_fn: impl FnOnce(I) -> anyhow::Result<AtomicOutput>,
+    ) -> anyhow::Result<AtomicExecID>
+    where
+        S: 'a + LockableState,
+    {
+        // Consume own input ID and retrieve the associated data
```

Comment:
How would you retrieve the associated data if it leaves in some other subnet?

Reply:
It is only about the local data which has been previously supplied to `init_atomic_exec`.